### PR TITLE
XOR the return address of all functions with the value of the global stack canary

### PR DESCRIPTION
**Pull request by StarMKWii**

### CODE REVIEW

File: payload/sp/security/Stack.S [hunk 1]
```diff
@@ -1,35 +1,49 @@
...
-    lwz       r0, 4(r31)
+// This function can be better optimized if we can find another register that is free for use
+.globl Stack_XORAlignedLinkRegisterSave
+Stack_XORAlignedLinkRegisterSave:
+    mtctr     r11
+
+    lis       r11, __stack_chk_guard@ha
+    lwz       r11, __stack_chk_guard@l(r11)
+    xor       r0, r0, r11
+    clrlwi    r0, r0, 1 // Null terminator
```
Comment (stblr) [inline payload/sp/security/Stack.S:11]: To improve performance, we could instead make sure that the msb of the canary itself is set.

File: payload/sp/security/Stack.c [hunk 2]
```diff
@@ -4,9 +4,32 @@
...
 typedef bool (*FindFunction)(const u32* start_address);
+typedef void (*LinkRegisterFunction)(void);
 
-extern void Stack_XORLinkRegisterRestore(void);
+typedef struct LinkRegisterPatch
+{
+    FindFunction find_prologue_function;
+    FindFunction find_epilogue_function;
+    FindFunction find_link_register_restore_function;
+    FindFunction find_link_register_save_function;
```
Comment (stblr) [inline payload/sp/security/Stack.c:14]: With how verbose this is getting, "link register" could maybe be abbreviated to just "lr". Also we use camelCase, so for instance `findLrSaveFunction`.

File: payload/sp/security/Stack.c [hunk 3]
```diff
@@ -24,108 +47,172 @@ __attribute__((noreturn)) void __stack_chk_fail(void)
...
 {
-    const u16 lr_save_instruction_high = 0x9001;
+    const u16 addi = 0x3821;
 
-    return (start_address[0] >> 16) == lr_save_instruction_high;
+    return start_address[0] == mtlr && (start_address[1] >> 16) == addi && start_address[2] == blr;
 }
 
-void Stack_DoLinkRegisterPatches(u32* start_address, u32* end_address)
+static const LinkRegisterPatch link_register_patches_array[2] =
```
Comment (stblr) [inline payload/sp/security/Stack.c:106]: Just `linkRegisterPatches` (or `lrPatches`).

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/sp/security/Stack.S (hunk 1 above)
Comment (StarMKWii) [reply payload/sp/security/Stack.S]: Addressed in commit [3535a85](https://github.com/stblr/mkw-sp/pull/280/commits/3535a85a673e6fcebfb5bb550cf34be59bd097f5).

File: payload/sp/security/Stack.c (hunk 2 above)
Comment (StarMKWii) [reply payload/sp/security/Stack.c]: Addressed in commit [7df0187](https://github.com/stblr/mkw-sp/pull/280/commits/7df0187cc678b51714bb875475a36a20e09fa00c).

File: payload/sp/security/Stack.c (hunk 3 above)
Comment (StarMKWii) [reply payload/sp/security/Stack.c]: Addressed in commit [7df0187](https://github.com/stblr/mkw-sp/pull/280/commits/7df0187cc678b51714bb875475a36a20e09fa00c).

**Review by stblr — APPROVED** [review]

**Pull request merged**